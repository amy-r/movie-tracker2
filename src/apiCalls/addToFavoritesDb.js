export const addToFavoritesDb = async (movieData, userId) => {
  // eslint-disable-next-line
  const favoriteData = { ...movieData, user_id: userId };
  const url = "api/users/favorites/new";
  try {
    const response = await fetch(url, {
      method: "POST",
      body: JSON.stringify(favoriteData),
      headers: { "Content-Type": "application/json" }
    });
    const favoriteId = await response.json();
    return favoriteId;
  } catch (error) {
    throw new Error("Error adding to favorites");
  }
};
