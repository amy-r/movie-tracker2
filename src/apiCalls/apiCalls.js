import apiKey from './apiKey';
import { movieCleaner } from './movieCleaner'

export const url = `https://api.themoviedb.org/3/discover/movie?api_key=${apiKey}&primary_release_date.gte=2018-02-27&primary_release_date.lte=2018-03-27`;

export const getMovies = async (url) => {
  try {
    const response = await fetch(url);
    const movies = await response.json();
    const cleanMovies = movieCleaner(movies.results);
    return cleanMovies;
  } catch (error) {
    throw new Error('Error getting movies');
  }
};

export const userLogin = async credentials => {
  const url = 'api/users';
  try {
    const response = await fetch(url, {
      method: 'POST',
      body: JSON.stringify(credentials),
      headers: { 'Content-Type': 'application/json' }
    });
    const user = await response.json();
    return user.data;
  } catch (error) {
    throw error;
  }
};

export const userSignup = async accountInfo => {
  const url = 'api/users/new';
  try {
    const response = await fetch(url, {
      method: 'POST',
      body: JSON.stringify(accountInfo),
      headers: { 'Content-Type': 'application/json' }
    });
    const userId = await response.json();
    return userId;
  } catch (error) {
    throw error;
  }
};
