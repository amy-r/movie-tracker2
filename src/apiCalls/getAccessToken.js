import { getTokenData } from "./getTokenData";

export const getAccessToken = async () => {
  const accessToken = sessionStorage.getItem('accessToken');
  const decodedToken = accessToken ? getTokenData(accessToken) : null;
  const expirationTime = decodedToken ? decodedToken.exp : null;
  const now = Math.trunc(Date.now() / 1000);

  // Token is expired or doesn't exist
  if (expirationTime && now >= expirationTime) {
    const apiRoot = process.env.REACT_APP_AUTH_API_URL;
    const url = apiRoot + "/auth/refresh";
    try {
      const response = await fetch(url, {
        method: "GET",
        headers: { "Content-Type": "application/json" },
        credentials: "include"
      });
      const tokens = await response.json();
      sessionStorage.setItem("accessToken", tokens.accessToken);
      return tokens.accessToken;
    } catch (error) {
      throw new Error("Error getting access token");
    }
  // Token is not expired
  } else if (now < expirationTime) {
    return accessToken;
  }
};