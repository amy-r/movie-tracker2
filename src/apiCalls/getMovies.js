const apiKey = process.env.REACT_APP_MDB_API_KEY

export const getMovies = async () => {
  const urlRoot = "https://api.themoviedb.org/3/movie/now_playing?";
  const urlKey = `api_key=${apiKey}`;
  const urlQuery = "&language=en-US&page=1";
  const url = `${urlRoot}${urlKey}${urlQuery}`;

  try {
    const response = await fetch(url);
    const movies = await response.json();
    return movies.results;
  } catch (error) {
    throw new Error("Error getting movies");
  }
};
