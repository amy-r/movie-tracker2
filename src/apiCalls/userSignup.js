export const userSignup = async accountInfo => {
  const apiRoot = process.env.REACT_APP_AUTH_API_URL;
  const url = apiRoot + "/users";
  try {
    const response = await fetch(url, {
      method: "POST",
      body: JSON.stringify(accountInfo),
      headers: { "Content-Type": "application/json" }
    });
    const userData = await response.json();
    const userId = userData.data.user.id;
    return userId;
  } catch (error) {
    throw new Error("Error signing up");
  }
};
