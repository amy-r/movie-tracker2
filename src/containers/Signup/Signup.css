.signup {
  width: 500px;
  height: 320px;
  padding: 30px;
  margin: 160px auto 0 auto;
  background-color: #1481BA;
}

input {
  margin: 15px auto;
}

@media screen and (max-width: 860px){

  .signup {
    height: 340px;
  }

}

@media screen and (max-width: 720px){

  .signup {
    width: 90%;
    padding: 10px;
  }

}

@media screen and (max-width: 500px){

  .signup {
    height: 350px;
  }

}

