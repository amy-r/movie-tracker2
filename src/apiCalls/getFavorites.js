/* eslint-disable camelcase */
import { getAccessToken } from "./getAccessToken";

export const getFavorites = async userId => {
  const token = await getAccessToken();
  if (!token) {
    return null;
  }
  const apiRoot = process.env.REACT_APP_AUTH_API_URL;
  const url = apiRoot + `/users/${userId}/favorites`;
  try {
    const response = await fetch(url, {
      method: "GET",
      headers: { 
        "Content-Type": "application/json",
        "Authorization": `Bearer ${token}`
      }
    });
    const favorites = await response.json();
    const cleanFavorites = favorites.data.favorites.map(favorite => {
      return {
        id: favorite.id,
        user_id: favorite.user_id,
        movie_id: parseInt(favorite.movie_id, 10),
        created_at: favorite.created_at
      };
    });
    return cleanFavorites;
  } catch (error) {
    throw new Error("Error getting favorites");
  }
};
