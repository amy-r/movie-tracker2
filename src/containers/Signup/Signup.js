import React, { Component } from 'react';
import './Signup.css';
import { userSignup } from '../../apiCalls/userSignup';
import { userLogin } from '../../apiCalls/userLogin';
import { validateUser } from '../../actions';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';

export class Signup extends Component {
  constructor() {
    super();

    this.state = {
      name: '',
      email: '',
      password: '',
      errorMessage: ''
    };
  }

  validateEmail = (email) => {
    // eslint-disable-next-line
    return /^([^\x00-\x20\x22\x28\x29\x2c\x2e\x3a-\x3c\x3e\x40\x5b-\x5d\x7f-\xff]+|\x22([^\x0d\x22\x5c\x80-\xff]|\x5c[\x00-\x7f])*\x22)(\x2e([^\x00-\x20\x22\x28\x29\x2c\x2e\x3a-\x3c\x3e\x40\x5b-\x5d\x7f-\xff]+|\x22([^\x0d\x22\x5c\x80-\xff]|\x5c[\x00-\x7f])*\x22))*\x40([^\x00-\x20\x22\x28\x29\x2c\x2e\x3a-\x3c\x3e\x40\x5b-\x5d\x7f-\xff]+|\x5b([^\x0d\x5b-\x5d\x80-\xff]|\x5c[\x00-\x7f])*\x5d)(\x2e([^\x00-\x20\x22\x28\x29\x2c\x2e\x3a-\x3c\x3e\x40\x5b-\x5d\x7f-\xff]+|\x5b([^\x0d\x5b-\x5d\x80-\xff]|\x5c[\x00-\x7f])*\x5d))*$/.test(email);
  }

  handleSubmit = async event => {
    event.preventDefault();
    let errorMessage = "";
    const validEmail = this.validateEmail(this.state.email);
    if (validEmail) {
      try {
        const signUpFetch = await userSignup(this.state);
        if (signUpFetch.error) {
          throw new Error("Error");
        } else {
          const user = await userLogin(this.state);
          this.props.validateUser(user);
          this.props.history.push("/");
        }
      } catch (error) {
        errorMessage = "Email address has already been used.";
      }
    } else {
      errorMessage = "Please enter a valid email address.";
    }
    this.setState({
      name: "",
      email: "",
      password: "",
      errorMessage: errorMessage
    });

  }

  handleChange = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  render() {
    const enableSubmit =
      this.state.name &&
      this.state.email &&
      this.state.password;
    return <div className="signup">
      <h4>Please enter your name, email address, <br />
      and password to create an account.</h4>
      <form onSubmit={this.handleSubmit}>
        <input 
          type="text" 
          name="name" 
          value={this.state.name} 
          placeholder="Name" 
          onChange={this.handleChange} />
        <input 
          type="text" 
          name="email" 
          value={this.state.email} 
          placeholder="Email address" 
          onChange={this.handleChange} />
        <input 
          type="password" 
          name="password" 
          value={this.state.password} 
          placeholder="Password" 
          onChange={this.handleChange} />
        <button 
          type="submit" 
          name="submit"
          disabled={!enableSubmit}>
            Submit
        </button>
      </form>
      <h2 className="errorMessage">{this.state.errorMessage}</h2>
    </div>;
  }
}

export const mapDispatchToProps = dispatch => ({
  validateUser: user => dispatch(validateUser(user))
});

export default withRouter(connect(null, mapDispatchToProps)(Signup));

Signup.propTypes = {
  validateUser: PropTypes.func,
  history: PropTypes.object
};
