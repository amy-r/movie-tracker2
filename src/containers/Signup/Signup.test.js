import React from 'react';
import { Signup, mapDispatchToProps } from './Signup';
import { shallow } from 'enzyme';
import { userSignup } from '../../apiCalls/userSignup';
import { userLogin } from '../../apiCalls/userLogin';

jest.mock('../../apiCalls/userSignup');
jest.mock('../../apiCalls/userLogin.js');

describe('Signup', () => {

  let wrapper;
  const event = { preventDefault: jest.fn() };
  const mockValidateUser = jest.fn();
  
  beforeEach(() => {
    wrapper = shallow(
      <Signup
        validateUser={mockValidateUser}
      />
    );
  });

  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('loads with an empty state', () => {
    expect(wrapper.state()).toEqual({
      name: '',
      email: '',
      password: '',
      errorMessage: ''
    });
  });

  it('calls validateEmail on handleSubmit', () => {
    wrapper.instance().validateEmail = jest.fn();
    const mockState = {
      name: 'me',
      email: 'me@gmail.com',
      password: 'password',
      errorMessage: ''
    };
    wrapper.setState({ email: 'me@gmail.com'});
    wrapper.instance().handleSubmit(event);
    expect(wrapper.instance().validateEmail).toHaveBeenCalledWith(mockState.email);
  });

  it('validates the email address input', () => {
    expect(wrapper.instance().validateEmail('me@gmail.com')).toEqual(true);
    expect(wrapper.instance().validateEmail('megmail.com')).toEqual(false);
    expect(wrapper.instance().validateEmail('megmailcom')).toEqual(false);
  })

  it('calls userSignup on handleSubmit when email address is valid', () => {
    wrapper.setState({
      name: 'me',
      email: 'me@gmail.com',
      password: 'password',
      errorMessage: ''
    });
    
    wrapper.instance().handleSubmit(event);
    expect(userSignup).toHaveBeenCalledWith(wrapper.state());
  });

  it('calls userLogin on handleSubmit when email address is valid', () => {
    wrapper.setState({
      name: "me",
      email: "me@gmail.com",
      password: "password",
      errorMessage: ""
    });

    wrapper.instance().handleSubmit(event);
    expect(userLogin).toHaveBeenCalledWith(wrapper.state());
  });

  it("calls validateUser on handleSubmit", () => {
    wrapper.instance().handleSubmit(event);
    expect(mockValidateUser).toHaveBeenCalled();
  });

  it('updates state with an error when email is invalid', () => {
    wrapper.setState({
      name: 'me',
      email: 'megmailcom',
      password: 'password',
      errorMessage: ''
    });
    const expected = "Please enter a valid email address.";
    wrapper.instance().handleSubmit(event);
    expect(wrapper.state('errorMessage')).toEqual(expected);
  });

  it.skip('updates state with an error when email is a duplicate', () => {
    wrapper.setState({
      name: "me",
      email: "tman2272@aol.com",
      password: "password",
      errorMessage: ""
    });
    const mockResponse = { error: "Key (email)=(tman2272@aol.com) already exists."};
    const expected = "Email address has already been used.";
    wrapper.instance().handleSubmit(event);
    expect(wrapper.state('errorMessage')).toEqual(expected);
  });

  it('resets the state and inputs when not redirected', () => {
    wrapper.setState({
      name: 'me',
      email: 'megmailcom',
      password: 'password',
      errorMessage: ''
    });
    const expected = {
      name: '',
      email: '',
      password: '',
      errorMessage: 'Please enter a valid email address.'
    };
    wrapper.instance().handleSubmit(event);
    expect(wrapper.state()).toEqual(expected);
  });

  it('updates the state as the user types', () => {
    const expected = {
      name: 'me',
      email: 'me@gmail.com',
      password: 'password',
      errorMessage: ''
    };
    const event1 = {
      target: {
        name: 'name',
        value: 'me'
      }
    };
    const event2 = {
      target: {
        name: 'email',
        value: 'me@gmail.com'
      }
    };
    const event3 = {
      target: {
        name: 'password',
        value: 'password'
      }
    };
    wrapper.instance().handleChange(event1);
    wrapper.instance().handleChange(event2);
    wrapper.instance().handleChange(event3);
    expect(wrapper.state()).toEqual(expected);
  });


  it.skip('updates state with an error message when credentials dont match', () => {
    const expected = "Email and password do not match."
    //mock throw error
    expect(wrapper.state('errorMessage').toEqual(expected));
  });

  it('disables submit button unless both fields contain data', () => {
    wrapper.setState({
      name: '',
      email: '',
      password: '',
      errorMessage: ''
    });
    expect(wrapper.find('button').prop('disabled')).toBe(true);
    wrapper.setState({
      name: 'me',
      email: '',
      password: '',
      errorMessage: ''
    });
    expect(wrapper.find('button').prop('disabled')).toBe(true);
    wrapper.setState({
      name: 'me',
      email: 'me@gmail.com',
      password: '',
      errorMessage: ''
    });
    expect(wrapper.find('button').prop('disabled')).toBe(true);
    wrapper.setState({
      name: 'me',
      email: 'me@gmail.com',
      password: 'password',
      errorMessage: ''
    });
    expect(wrapper.find('button').prop('disabled')).toBe(false);
  });

  it('should call dispatch function when using mapDispatchToProps', () => {
    const mockDispatch = jest.fn();
    const mapped = mapDispatchToProps(mockDispatch);
    mapped.validateUser();
    expect(mockDispatch).toHaveBeenCalled();
  });
});