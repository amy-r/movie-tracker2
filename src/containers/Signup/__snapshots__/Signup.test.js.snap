// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Signup
    validateUser={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h4>
          Please enter your name, email address, 
          <br />
          and password to create an account.
</h4>,
        <form
          onSubmit={[Function]}
>
          <input
                    name="name"
                    onChange={[Function]}
                    placeholder="Name"
                    type="text"
                    value=""
          />
          <input
                    name="email"
                    onChange={[Function]}
                    placeholder="Email address"
                    type="text"
                    value=""
          />
          <input
                    name="password"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password"
                    value=""
          />
          <button
                    disabled={true}
                    name="submit"
                    type="submit"
          >
                    Submit
          </button>
</form>,
        <h2
          className="errorMessage"
/>,
      ],
      "className": "signup",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Please enter your name, email address, ",
            <br />,
            "and password to create an account.",
          ],
        },
        "ref": null,
        "rendered": Array [
          "Please enter your name, email address, ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          "and password to create an account.",
        ],
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              name="name"
              onChange={[Function]}
              placeholder="Name"
              type="text"
              value=""
/>,
            <input
              name="email"
              onChange={[Function]}
              placeholder="Email address"
              type="text"
              value=""
/>,
            <input
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
              value=""
/>,
            <button
              disabled={true}
              name="submit"
              type="submit"
>
              Submit
</button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "name",
              "onChange": [Function],
              "placeholder": "Name",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "email",
              "onChange": [Function],
              "placeholder": "Email address",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "password",
              "onChange": [Function],
              "placeholder": "Password",
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Submit",
              "disabled": true,
              "name": "submit",
              "type": "submit",
            },
            "ref": null,
            "rendered": "Submit",
            "type": "button",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
          "className": "errorMessage",
        },
        "ref": null,
        "rendered": "",
        "type": "h2",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h4>
            Please enter your name, email address, 
            <br />
            and password to create an account.
</h4>,
          <form
            onSubmit={[Function]}
>
            <input
                        name="name"
                        onChange={[Function]}
                        placeholder="Name"
                        type="text"
                        value=""
            />
            <input
                        name="email"
                        onChange={[Function]}
                        placeholder="Email address"
                        type="text"
                        value=""
            />
            <input
                        name="password"
                        onChange={[Function]}
                        placeholder="Password"
                        type="password"
                        value=""
            />
            <button
                        disabled={true}
                        name="submit"
                        type="submit"
            >
                        Submit
            </button>
</form>,
          <h2
            className="errorMessage"
/>,
        ],
        "className": "signup",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Please enter your name, email address, ",
              <br />,
              "and password to create an account.",
            ],
          },
          "ref": null,
          "rendered": Array [
            "Please enter your name, email address, ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            "and password to create an account.",
          ],
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                name="name"
                onChange={[Function]}
                placeholder="Name"
                type="text"
                value=""
/>,
              <input
                name="email"
                onChange={[Function]}
                placeholder="Email address"
                type="text"
                value=""
/>,
              <input
                name="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
/>,
              <button
                disabled={true}
                name="submit"
                type="submit"
>
                Submit
</button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "name",
                "onChange": [Function],
                "placeholder": "Name",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "email",
                "onChange": [Function],
                "placeholder": "Email address",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "name": "password",
                "onChange": [Function],
                "placeholder": "Password",
                "type": "password",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Submit",
                "disabled": true,
                "name": "submit",
                "type": "submit",
              },
              "ref": null,
              "rendered": "Submit",
              "type": "button",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
            "className": "errorMessage",
          },
          "ref": null,
          "rendered": "",
          "type": "h2",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
